Handle credit validation errors

sauloperez opened the pull request:
#### What? Why?

Closes https://github.com/openfoodfoundation/openfoodnetwork/issues/5449

Note this is based on #5806.

I'm not particularly happy about having to come down to the `attr_accessor` trick to skip a callback but this prevents this fix from growing exponentially on LOCs and avoids the risk of making the wrong assumptions. What I like about it is that `skip_source_validation` is explicit enough to pave the way for extracting specific per use-case service objects in the future.

#### What should we test?

We need to cover all payment scenarios:

* Issue a partial credit with each supported payment gateway
* Issue a void operation with each supported payment gateway
* Issue a capture operation with each supported payment gateway
* Place a regular order as a consumer
* Place an order as an admin

When performing these actions as admin, the payment total should be accurate, taking into account any payment fees.

#### Release notes

Handle credit and other payment actions whose credit card is now expired. Stripe kept track of them and we weren't
Changelog Category: Fixed

luisramos0 commented on app/models/spree/payment.rb:
```diff
@@ -160,6 +160,14 @@ def validate_source
+    def valid_source?
+      if source.is_a?(Spree::Payment)
+        true
+      else # top-most source
```
this is very hard for me to understand even if I read the commit message and if know the context inside out...
what if we extract this little method to a little service and add a nice comment to explain?

sauloperez replied:
I think I can manage with a good method doc. This won't be pretty as we are approaching callbacks nastiness but bear with me.

luisramos0 commented on spec/models/spree/payment_spec.rb:
```diff
@@ -1,7 +1,691 @@
+
+        context "when response is successful" do
+          it "should create an offsetting payment" do
+            Spree::Payment.should_receive(:create!)
```
this syntax can be modernized right? expect .... to receive.

sauloperez replied:
yeah, it should. Good catch.

luisramos0 commented on spec/models/spree/payment_spec.rb:
```diff
@@ -401,6 +401,25 @@
+              ActiveMerchant::Billing::Response.new(true, "Yay!")
+            end
+
+            it 'lets the new payment to be saved' do
```
I think the correct English is "it 'lets the new payment be saved'" or ""it 'allows the new payment to be saved'""

luisramos0 commented on app/models/spree/payment.rb:
```diff
@@ -151,7 +163,7 @@ def revoke_adjustment_eligibility
 
     def validate_source
-      if source && !source.valid?
+      if source && !skip_source_validation && !source.valid?
```
I know you only added a variable here :-)
I wonder if this is not easier and equal if written like:
```ruby
def validate_source
   return true if !source || skip_source_validation || source.valid?

   source.errors.each do |field, error|
     field_name = I18n.t("activerecord.attributes.#{source.class.to_s.underscore}.#{field}")
     errors.add(Spree.t(source.class.to_s.demodulize.underscore), "#{field_name} #{error}")
    end
    false
end

```

sauloperez replied:
ok, I gave it a try and then I thought. Wait and see if someone mentions it :joy:

luisramos0 commented on app/models/spree/payment.rb:
```diff
@@ -176,8 +188,12 @@ def create_payment_profile
       gateway_error e
     end
 
+    # Inherited from Spree, makes newly entered payments invalidate previously
```
I dont think we need to tell our story here. I'd remove "Inherited from Spree,".

sauloperez replied:
ok. What's important is the rest.

luisramos0 commented on app/models/spree/payment.rb:
```diff
@@ -32,7 +32,19 @@ class Payment < ActiveRecord::Base
     # invalidate previously entered payments
     after_create :invalidate_old_payments
 
+    # Prevents #validate_source to end up in a chain of validations, from
```
I found this comment quite complicated. Let me try:

First I'd describe what the attribute does and then say where it is used and why:

```ruby
# Skips the validation of the source (for example, credit card) of the payment.
# 
# This is used in refunds as the validation of the card can fail but the refund can go through, we trust the payment 
#    gateway in these cases. For example, Stripe is accepting refunds with source cards that were valid when
#    the payment was placed but are now expired, and we consider them invalid.

```
I think that's it here.

I dont think the chain of validations is related to this refund/skip_source_validation, is it? I think it's related to the invalidation of previous payments that you have wisely changed below from invalidate! to update_column :+1: 
There (line 191), I'd add this comment:
```ruby
# Using update_column skips validations and so it skips validate_source. As we are just invalidating
#     past payments here, we don't want to validate all of them at this stage.
```

sauloperez replied:
Thanks, that looks better. 

> I dont think the chain of validations is related to this refund/skip_source_validation, is it?

It was happening due to both. Every source was validating its source up until the credit card.

sauloperez commented:
@luisramos0 I addressed your comments plus a commit that handles `#refund!`.

Matt-Yorkley commented on app/models/spree/payment.rb:
```diff
@@ -176,8 +184,14 @@ def create_payment_profile
+    # is used by the gateway.
     def invalidate_old_payments
-      order.payments.with_state('checkout').where("id != ?", id).each(&:invalidate!)
+      order.payments.with_state('checkout').where("id != ?", id).each do |payment|
```
I think we can start using `#not` in Rails 4 for these "where not equals" AR scopes. Not blocking, just mentioning it...

sauloperez replied:
wooow! I totally forgot Rails 4.0 introduced [#not](https://apidock.com/rails/v4.0.2/ActiveRecord/QueryMethods/WhereChain/not)!

sauloperez commented:
As expected, chaning from `#invalidate!` to `#update_column` doesn't play with the order total calculation in the build :see_no_evil:

sauloperez commented:
I thought this would take days but I think I managed to solve it. I also addressed your comment @Matt-Yorkley . ProTip: You need a where because comes out of an association is a ` #ActiveRecord::AssociationRelation` and not a `WhereChain`. I guess the latter substitutes the `ActiveRecord::Relation` in Rails 4.

filipefurtad0 commented:
Hi @sauloperez and @luisramos0,

The PR introduces an error 500 on /admin/enterprises/

Please see:
https://openfoodnetwork.slack.com/archives/CEF14NU3V/p1595523810421200

sauloperez commented:
Got it. The migration  https://github.com/openfoodfoundation/openfoodnetwork/blob/master/db/migrate/20200616162646_move_all_calculators_outside_the_spree_namespace.rb was applied in that DB but this branch's code still has the calculators within the Spree namespace.

sauloperez commented:
I forced push to the get the latest changes of https://github.com/openfoodfoundation/openfoodnetwork/pull/5806.

sauloperez commented:
@filipefurtad0 it's solved now.

filipefurtad0 commented:
Hey @sauloperez ,

I've reversed the order, and numbered the test cases:

i) Place regular orders, as a consumer
ii) Place orders as admin
iii) Issue a partial credit with each supported payment gateway
iv) Issue a void operation with each supported payment gateway
v) Issue a capture operation with each supported payment gateway

There is quite a lot going on here, so let's go, vaya vayita! :-)


Set different enterprise fees, shipping fees, transaction fees. The shop charges GST. Placed 6 orders, with different payment methods in the shopfront. All payment methods have the same transaction fee.

**i) Placing 6 orders, with these payment methods:**

1) Paypal
2) Cash
3) Stripe-SCA (paid with 3D card)
4) Stripe-SCA (paid with regular)
5) Stripe-Connect
6) Pin payments

The value of the orders should be and is the same, as checked on confirmation e-mails and under /orders.


**ii) Placing the same 6 orders, on the backoffice:**

1) Paypal - No Payment Via Admin Backend
2) Cash - OK
3) Stripe-SCA (paid with 3D card) - not supported in the Backoffice
4) Stripe-SCA (paid with regular)  - OK
5) Stripe-Connect - OK
6) Pin payments – not supported in the back office - #5790

Issue “Backoffice orders: "Balance due" state after payments with a transaction fee #5574” was observed; Used the described workaround and obtained the same values as those obtained in shopfront orders.


**iii) Issue a partial credit with each supported payment gateway**

Removed 5 items and recalculated fees. Proceeded to Payments:

1) Paypal - issuing credit is not supported

![image](https://user-images.githubusercontent.com/49817236/88391207-a33fe300-cdb1-11ea-97a3-0119c67768ba.png)

2) Cash - issuing credit is not supported

![image](https://user-images.githubusercontent.com/49817236/88391357-f154e680-cdb1-11ea-84be-59b7d7573bca.png)

3) Stripe-SCA (paid with 3D card) - Issuing the credit works. Known issue [#3490](When crediting an order the transaction fee is applied when it shouldn't)

![image](https://user-images.githubusercontent.com/49817236/88391510-3e38bd00-cdb2-11ea-9871-6da4de03b259.png)

![image](https://user-images.githubusercontent.com/49817236/88391445-18131d00-cdb2-11ea-9ec6-2196f091d504.png)

![image](https://user-images.githubusercontent.com/49817236/88391538-4b55ac00-cdb2-11ea-9eea-98ee9956de01.png)


Refund appears correctly in Stripe Dashboard:

![image](https://user-images.githubusercontent.com/49817236/88392591-26facf00-cdb4-11ea-9800-6ed6b3ab9a57.png)


4) Stripe-SCA (paid with regular) - same as 3)
5) Stripe-Connect - same as 3)
6) Pin payments – issuing is not supported:

![image](https://user-images.githubusercontent.com/49817236/88391880-ef3f5780-cdb2-11ea-893b-f6bb0741874f.png)




**iv) Issue a void operation with each supported payment gateway:**

1) Paypal - No Payment Via Admin Backend - not supported.
2) Cash - "Voiding payment does not remove transaction fee #3584"
3) Stripe-SCA (paid with 3D card) - If one tries to void the payments on the previously edited order, one gets the errors:

- voiding the refund:

![image](https://user-images.githubusercontent.com/49817236/88392785-8527b200-cdb4-11ea-8492-37f6c86042e5.png)

- voiding the initial payment:

![image](https://user-images.githubusercontent.com/49817236/88392686-5c072180-cdb4-11ea-84d1-58eaf722debb.png)


- Voiding a Stripe payment from an order which was not edited works fine.


4) Stripe-SCA (paid with regular) - same as 3)

5) Stripe-Connect - Voiding the initial payment from the order works. Attempting to void the refund, after refunding the whole payment, results in this error:

![image](https://user-images.githubusercontent.com/49817236/88394402-4d6e3980-cdb7-11ea-9d8e-3bf96d912dae.png)


On Stripe dashboard it can be seen, that a partially refunded order goes from this state:

![image](https://user-images.githubusercontent.com/49817236/88395792-860f1280-cdb9-11ea-8c34-5d55c672ec9f.png)


Into:

![image](https://user-images.githubusercontent.com/49817236/88395980-ca9aae00-cdb9-11ea-8574-3bff233e2c55.png)


6) Voiding Pin payments – is not supported in the backoffice:

![image](https://user-images.githubusercontent.com/49817236/88393007-e94a7600-cdb4-11ea-9837-1df43c3e19e2.png)



**v) Issue a capture operation with each supported payment gateway**

I am not 100% sure about this operation @sauloperez . Do you mean this (below) the tick, under /orders? I thought this was only possible for cash payments...

![image](https://user-images.githubusercontent.com/49817236/88393446-89a09a80-cdb5-11ea-9246-2f59933d01c4.png)

2) Cash - Capturing payment works


So in summary:

- I think there is still an issue with refunds on Stripe-SCA. It should be possible to refund the initial amount, without getting the error. This is perhaps worth opening a new issue.

- for Stripe Connect (for which the issue was opened) all looks good.

- The PR touches several previously known issues. See epic [5580](https://github.com/openfoodfoundation/openfoodnetwork/issues/5580). 

I would say this is ready to go, but I'm especially grateful for feedback here. This is quite extensive and touches many corners. Moving to ready to go, but adding the feedback label. Thank you!

luisramos0 commented:
Hey @filipefurtad0 thanks a lot for the testing.
The testing looks good. Re the StripeSCA, I think the refund worked well:
![image](https://user-images.githubusercontent.com/1640378/88397575-09316800-cdbc-11ea-9965-3b8465506052.png)

I think you tried to void the refund and then void the original payment of an order with a partial refund. I think it's good it failed. After you issue a partial refund, foer example, payed 100 and refunded 50, none of these two movements should be voided. I think all the user can do after this is to issue another partial refund.

I think what also needs to be tested here are the checkout retrial scenarios. I will have a look at #5816 and #5818 so we can continue the conversation there.

So, I will merge this PR now and look at #5816 and #5818 with and without this PR to see if anything has changed :+1: 
Thanks Filipe!

filipefurtad0 commented:
Hey @luisramos0,

Yes, the partial refund works well for both Stripe methods. The subsequent full refund fails for Stripe-SCA (test cases iv) 3 and 4) but works nicely for Stripe-Connect (see test case iv) 5).

If not allowing for a full refund after a partial one is the expected behavior than all is good think, as eventually Stripe-Connect will be discontinued.

I checked anyway for this difference between the two payment methods and it is not introduced by this PR and it does not seem to relate to transaction fees.

luisramos0 commented:
oh, nice finding. I didnt know that would work in StripeConnect. That's basically lost funcionality from StripeConnect to StripeSCA. We should create a specific issue for it and then decide what to do.

The functitonality is "full refund after partial refund", and I think you mean the user presses VOID on the original payment, right?
I wonder if VOID on the partial refund also works on StripeConnect. Voiding a partial refund should work in StripeSCA, I think that could another separate issue to be looked at.
These are straight forward things to do in stripeSCA, when we have the issues, we can decide if we do them now or later.

filipefurtad0 commented:
Hi @luisramos0 ,

> I wonder if VOID on the partial refund also works on StripeConnect. Voiding a partial refund should work in StripeSCA, I think
> that could be another separate issue to be looked at.

Voiding partial refunds do not work in neither Stripe-Connect or -SCA: opened issue #5830

luisramos0 commented:
Hello Filipe! Thanks for confirming that :+1:

sauloperez commented:
Thanks a lot for this in-depth testing @filipefurtad0 . We are now on a much better position to fix these things :muscle:
